R/binary_beta.R

```diff
@@ -0,0 +1,2567 @@
+    dplyr::collect()
+
+  # basic sanity checks on collected data
+  # when ps is <phip_data>, its already validated
+  if (anyNA(dat_small$sample_id) || anyNA(dat_small$peptide_id)) {
+    .ph_abort(
+      "`ps` contains missing values in `sample_id` and/or
+      `peptide_id`."
+    )
+  }
+
+  # duplicates will break pivot_wider (or create list-cols); require uniqueness
+  if (anyDuplicated(dat_small[, c("sample_id", "peptide_id")]) > 0L) {
+    .ph_abort(
+      "found duplicated (sample_id, peptide_id) pairs in `ps`."
+    )
+  }
+
+  # replace nas with 0 in abundance column
+  dat_small[[value_col]][is.na(dat_small[[value_col]])] <- 0
+
+  # ensure abundance is numeric
+  if (!is.numeric(dat_small[[value_col]])) {
+    .ph_abort(
+      paste0("`", value_col, "` must be numeric after collect().")
+    )
+  }
+
+.ph_log_info("pivoting to wide abundance matrix in r.")
+
+  wide_df <- dat_small |>
+    tidyr::pivot_wider(
+      id_cols     = sample_id,
+      names_from  = peptide_id,
+      values_from = !!value_sym,
+      values_fill = 0
+    )
+
+if (!"sample_id" %in% names(wide_df)) {
+  .ph_abort("failed to construct wide abundance table (no `sample_id`).")
+}
+
+  mat <- wide_df |>
+    tibble::column_to_rownames("sample_id") |>
+    as.matrix()
+
+  if (nrow(mat) == 0L || ncol(mat) == 0L) {
+    .ph_abort(
+      "abundance matrix is empty after reshaping. check filters and
+      `value_col`."
+    )
+  }
+
+.ph_log_info(
+  paste0("abundance matrix has ", nrow(mat), " samples and ",
+         ncol(mat), " features after preprocessing.")
+)
+
+  # ---------------------------------------------------------------------------
+  # 3) normalization
+  # ---------------------------------------------------------------------------
+  if (identical(method_normalization, "auto")) {
+    vals <- mat[!is.na(mat)]
+    is_binary_data <- length(vals) > 0L && all(vals == 0 | vals == 1)
+    method_normalization <- if (is_binary_data) "none" else "relative"
+  .ph_log_info(
+    paste0("auto normalization selected -> using ", method_normalization)
+  )
+}
+
+  norm_mat <- switch(
+    method_normalization,
+    "none" = mat,
+    "relative" = {
+      rs <- rowSums(mat, na.rm = TRUE)
+      rs[rs == 0] <- 1
+      mat / rs
+    },
+    "hellinger" = {
+      rs <- rowSums(mat, na.rm = TRUE)
+      rs[rs == 0] <- 1
+      sqrt(mat / rs)
+    },
+    "log" = log1p(mat)
+  )
+
+  # ---------------------------------------------------------------------------
+  # 4) distance computation
+  # ---------------------------------------------------------------------------
+  if (dist_method == "chebyshev") dist_method <- "maximum"
+
+.ph_log_info(
+  paste0("computing distance: ", dist_method)
+)
+
+  dist_obj <- NULL
+
+  pd_methods <- c(
+    "bray", "euclidean", "minkowski", "manhattan", "canberra",
+    "binary", "maximum", "cosine"
+  )
+
+  use_pd <- dist_method %in% pd_methods
+
+  if (use_pd) {
+    if (!rlang::is_installed("parallelDist")) {
```

Comment:
Some of the `parallelDist` names are also valid for `vegan::vegdist` (like `bray`, `euclidean`, `manhattan`, etc.) and these should go to `vegan` even if `parallelDist` is not installed (instead of aborting).